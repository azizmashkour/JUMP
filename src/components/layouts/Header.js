import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { withTranslation } from 'react-i18next';
import BrowserLanguage from '../../utils/BrowserLanguage';
import Select from 'react-select';
import { options, langs } from '../../configs/options';

const Header = ({lang, i18n, t}) => {
  const [state, setState] = useState({
    defaultLang: {
      label: i18n.language === 'en' ?  langs.en: langs.fr,
      value: i18n.language === 'en' ? langs.en : langs.fr,
    }
  });

  const changeLang = (lang) => {
    if (!lang) {
      const tempLang = BrowserLanguage.getDefaultLanguage();
      lang = tempLang === 'en' ? 'fr' : 'en';
    }
    setState({ defaultLang: lang });

    if (BrowserLanguage.setLanguage(lang.value)) {
      // Reload page if browser support localStorage
      window.location.reload(); // Relaod app after langue change
    } else {
      // Change language at runtime if localStorage not found
      i18n.changeLanguage(lang); // Need for change language at runtime
    }
  };

  return (
    <div className="header-container">
      <Select
        defaultValue={state.defaultLang}
        options={options}
        value={state.defaultLang}
        onChange={changeLang}
        className="App-Select"
        isSearchable={false}
      />
    </div>
  );
};

Header.propTypes = {
  t: PropTypes.func.isRequired,
  i18n: PropTypes.shape({
    defaultNS: PropTypes.string,
    changeLanguage: PropTypes.func,
  }).isRequired,
};

export default withTranslation()(Header);
