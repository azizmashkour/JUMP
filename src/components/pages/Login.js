import React, { useState } from 'react';
import { withTranslation } from 'react-i18next';
import { options } from '../../configs/options';

const Login = ({lang, i18n, t}) => {
  const [state, setState] = useState({ lang: options[0], });

  const formRef = React.createRef();
  const usernameRef = React.createRef();
  const passwordRef = React.createRef();

  // Show input error message
  const showError = (input, message) => {
    const formControl = input.parentElement;
    formControl.className = "form-control error";
    const small = formControl.querySelector("small");
    small.innerText = message;
  }

  // Show success outline
  const showSuccess = (input) => {
    const formControl = input.parentElement;
    formControl.className = "form-control success";
  }

  // To make the first letter in the error message capital
  const getFieldName =(input) => {
    return input.id.charAt(0).toUpperCase() + input.id.slice(1);
  }

  // Check Input Length
  const checkLength =(input, min, max) => {
    if (input.current.value.length < min) {
      showError(input.current, `${getFieldName(input.current)} must be at least ${min}`);
    } else if (input.current.value.length > max) {
      showError(input.current, `${getFieldName(input.current)} must be at less than ${max}`);
    } else {
      showSuccess(input.current);
      this.props.history.replace('/');
    }
  }

  // Check required fields
  const checkRequired = (inputArr) => {
    inputArr.forEach((input) => {
      if (input.current.value.trim() === "") {
        showError(input.current, `${getFieldName(input.current)} is required`);
      } else {
        showSuccess(input.current);
      }
    });
  }

  const handleInputChange = (attr) => {
    attr === 'username' ? checkLength(usernameRef, 3, 15) : checkLength(passwordRef, 6, 20);
  }

  const submitForm = (event) => {
    event.preventDefault();
    checkRequired([usernameRef, passwordRef]);
  }

  return (
    <div>
    <h1 className="Lab__title">Application JUMP</h1>
    <div className="container">
        <form id="form" className="form" ref={formRef}>
          <h2>Veuillez vous authentifier</h2>
          <div className="form-control">
            <label for="username">Utilisateur</label>
            <input type="text" id="username" placeholder="Enter username" ref={usernameRef} onChange={(e) =>handleInputChange('username')}/>
            <small>Error Message</small>
          </div>
          <div className="form-control">
            <label for="password">Mot de passe</label>
            <input type="password" id="password" placeholder="Enter password" ref={passwordRef} onChange={(e) =>handleInputChange('password')}/>
            <small>Error Message</small>
          </div>
          <button type="submit" onClick={(e) =>submitForm(e)}>Submit</button>
        </form>
    </div>
    </div>
  );
};

export default withTranslation()(Login);
