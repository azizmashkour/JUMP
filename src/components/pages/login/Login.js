import React from 'react';
import PropTypes from 'prop-types';
import { withTranslation } from 'react-i18next';
import Header from '../../layouts/Header';

// this component fake the login authentication form for any type of user
// matching the validation criteria: username and password constraints.

const Login = (props) => {
  // Create references for Login form tag, username and password inputs tags
  const formRef = React.createRef();
  const usernameRef = React.createRef();
  const passwordRef = React.createRef();

  // Show input error message
  const showError = (input, message) => {
    const formControl = input.parentElement;
    formControl.className = "form-control error";
    const small = formControl.querySelector("small");
    small.innerText = message;
  }

  // Show success outline
  const showSuccess = (input) => {
    const formControl = input.parentElement;
    formControl.className = "form-control success";
  }

  // To make the first letter in the error message capital
  const getFieldName =(input) => {
    return input.id.charAt(0).toUpperCase() + input.id.slice(1);
  }

  // Check Input Length
  const checkLength =(input, min, max) => {
    if (input.current.value.length < min) {
      showError(input.current, `${getFieldName(input.current)} must be at least ${min}`);
      return false;
    } else if (input.current.value.length > max) {
      showError(input.current, `${getFieldName(input.current)} must be at less than ${max}`);
      return false;
    } else {
      showSuccess(input.current);
      return true;
    }
  }

  // Check required fields
  const checkRequired = (inputArr) => {
    inputArr.forEach((input) => {
      if (input.current.value.trim() === "") {
        showError(input.current, `${getFieldName(input.current)} is required`);
      } else {
        showSuccess(input.current);
      }
    });
  }

  // Check validation on form inputs change
  const handleInputChange = (attr) => {
    attr === 'username' ? checkLength(usernameRef, 3, 15) : checkLength(passwordRef, 6, 20);
  }

  // Only submit successfully validated form then redirect to next page
  const submitForm = (event) => {
    event.preventDefault();
    checkRequired([usernameRef, passwordRef]);
    const usernameChecker = checkLength(usernameRef, 3, 15);
    const passwordChecker = checkLength(passwordRef, 6, 20);
    if (usernameChecker && passwordChecker) {
      props.history.replace('/clients');
    }
  }

  const  { t } = props;

  return (
    <div>
      <Header />
      <h1 className="Lab__title">{t('login.jump_application')}</h1>
      <div className="container">
        <form id="form" className="form" ref={formRef}>
          <h2>{t('login.please_authenticate')}</h2>
          <div className="form-control">
            <label htmlFor="username">{t('login.user')}</label>
            <input
              autoComplete="off"
              type="text"
              id="username"
              name="username"
              placeholder={t('login.enter_username')}
              ref={usernameRef}
              onChange={(e)=>handleInputChange('username')}
            />
            <small>{t('login.error_message')}</small>
          </div>
          <div className="form-control">
            <label htmlFor="password">{t('login.password')}</label>
            <input
              autoComplete="off"
              type="password"
              id="password"
              name="password"
              placeholder={t('login.enter_password')}
              ref={passwordRef}
              onChange={(e)=>handleInputChange('password')}
            />
            <small>{t('login.error_message')}</small>
          </div>
          <button type="submit" onClick={(e) =>submitForm(e)}>{t('login.submit')}</button>
        </form>
      </div>
    </div>
  );
};

Login.propTypes = {
  t: PropTypes.func.isRequired,
};

export default withTranslation()(Login);
