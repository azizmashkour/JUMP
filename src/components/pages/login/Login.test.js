import React from "react";
import { shallow } from "enzyme";
import Login from "./Login";
import Header from '../../layouts/Header';

describe("Login", () => {
  const mockSubmit = jest.fn();
  const wrapper = shallow(<Login submit={mockSubmit} />);

  it("should display Header component", () => {
    expect(wrapper.find(Header)).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  describe("handleChange", () => {
    const inputs = [
      { name: 'username', value: 'azizmashkour', id: 'username' },
      { name: 'password', value: 'Test#123', id: 'password' },
    ];

    const event = {
      preventDefault() {},
      target: { value: "the value" }
    };

    inputs.forEach(({ name, value }) => {
      it(`calls handleChange for ${name}`, () => {
        // console.log('namenamename', name, `#${name}`);
        // const input = wrapper.find(`#${name}`);
        // console.log('DEBUGGING::input finding word', input.debug());
        // console.log('input finding word LENGHTS', input.length);

        // const input = wrapper.find({ name }).first();
        // expect(input.exists()).toEqual(true)
        //
        // input.simulate('change', { target: { value } });
        //
        // console.log(input.instance());
        //
        // expect(handleChange).toBeCalled();
      });
    });

    // wrapper.instance().handleChange(event);
    //
    // expect(wrapper.state()).toEqual(expected);
  })

})
