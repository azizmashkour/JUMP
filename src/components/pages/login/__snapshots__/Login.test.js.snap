// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should display Header component 1`] = `ShallowWrapper {}`;
